### ISSUE TITLE
How to wait for all configuration data to be available?

### ISSUE THREAD

**Issue by jimboca** [gwww/elkm1#9]
I'm moving my ISY Elk Polyglot NodeServer https://github.com/jimboca/udi-poly-elk over from pyelk to this library, but I need to know when all information is available, and would like to process all Area's before Zones. When I do this with elkm1 the data is not all available yet.  Meaning I do this,
```
        elk = Elk({'url': url})
        elk.connect()
        thread = Thread(name='ELK_RUN',target=elk.run)
        for area_number in range(7):
            LOG.debug('area {}'.format(elk.areas[area_number]))
            elk.areas[area_number].add_callback(call_me)
```
Which shows them all not initialized
```
area 0 'Area-001' armed_status:None arm_up_state:None alarm_state:None alarm_memory:None is_exit:False timer1:0 timer2:0
area 1 'Area-002' armed_status:None arm_up_state:None alarm_state:None alarm_memory:None is_exit:False timer1:0 timer2:0
area 2 'Area-003' armed_status:None arm_up_state:None alarm_state:None alarm_memory:None is_exit:False timer1:0 timer2:0
area 3 'Area-004' armed_status:None arm_up_state:None alarm_state:None alarm_memory:None is_exit:False timer1:0 timer2:0
area 4 'Area-005' armed_status:None arm_up_state:None alarm_state:None alarm_memory:None is_exit:False timer1:0 timer2:0
area 5 'Area-006' armed_status:None arm_up_state:None alarm_state:None alarm_memory:None is_exit:False timer1:0 timer2:0
area 6 'Area-007' armed_status:None arm_up_state:None alarm_state:None alarm_memory:None is_exit:False timer1:0 timer2:0
```

The callback looks like:
```
def call_me(element, changeset):
       print('el={} cs={}'.format(element,changeset))
```
So, sometime later I get the real information about the Area:
```
el=0 'Home' armed_status:0 arm_up_state:0 alarm_state:0 alarm_memory:None is_exit:False timer1:0 timer2:0 cs={'name': 'Home'}
```

I can track this all internally to make sure all the Area is initialized before working on the Zones in that area, but was hoping there was an easier way?

Thanks

**Comment by gwww** [comment]
Your are right, there is nothing supported in the library as is. I just took a quick peek at the code (the sync() methods) and there's no easy way to mod them to tag when a sync is complete. I'll think about it a bit, perhaps the mystery of sleep will create an idea.

**Comment by jimboca** [comment]
Much appreciated, I'm willing to help with any coding changes if you dream up an idea.

**Comment by gwww** [comment]
One thing you could do, while not "perfect", is to time how long it takes to sync on average. Then set a "done" timer to wake up the rest of your app. Here's an example of setting a timer...
https://github.com/gwww/elkm1/blob/master/elkm1_lib/elk.py#L100

**Comment by jimboca** [comment]
Thanks, I'll think about that, but I was adding callback's that would create the objects the first time they were called which I think will work, but I'm having an issue trying to get the asyncio working because these node servers are run in a thread, so initially I get the error:
```
  File "/home/pi/development/udi-poly-elk/nodes/Controller.py", line 113, in discover
    self.elk = Elk(config)
  File "/home/pi/.local/lib/python3.6/site-packages/elkm1_lib/elk.py", line 21, in __init__
    self.loop = loop if loop else asyncio.get_event_loop()
  File "/usr/local/lib/python3.6/asyncio/events.py", line 694, in get_event_loop
    return get_event_loop_policy().get_event_loop()
  File "/usr/local/lib/python3.6/asyncio/events.py", line 602, in get_event_loop
    % threading.current_thread().name)
RuntimeError: There is no current event loop in thread 'NodeServer'.
```

I have no experience with python asyncio, so trying to research I found some say to do this in the thread:
```
        asyncio.set_event_loop(asyncio.new_event_loop())
```
and after adding that, I no longer get the error, but I don't see anything happening with full log debug enabled.

**Comment by gwww** [comment]
Here is the `bin/simple` program updated. I used a 5 second timer... change as needed.
```
#!/usr/bin/env python

from elkm1_lib import Elk
import logging
import os

LOG = logging.getLogger(__name__)
config_complete_timer = None

def main():
    logging.basicConfig(level=logging.DEBUG, format='%(message)s')
    try:
        url = os.environ.get('ELKM1_URL')
        if not url:
            print("Specify url to connect to in ELKM1_URL environment variable")
            exit(0)

        elk = Elk({'url': url})
        elk.connect()
        config_complete_timer = elk.loop.call_later(5, config_complete)
        elk.run()
    except KeyboardInterrupt:
        exit(0)

def config_complete():
    print("Config complete... do whatever you need to do here...")

if __name__ == "__main__":
    main()
```

**Comment by jimboca** [comment]
Thanks, I appreciate that.  I'll give it a try, just worried not knowing
for sure that the sync has completed.   What would you think about passing
in an Event which could be set when sync is complete?

Also, any idea on how to use this in a thread since the thread doesn't have
an event loop?  I've tried many options but can't seem to figure it out.

**Comment by gwww** [comment]
I would have an event if that were possible. Because synchronizing with the panel is done  asynchronously there's no way to easily know when its complete. 

I'm not a thread expert but here's one idea. When the sync is "complete" (the timer expires), start the thread with a "complete" config. Not sure that fits with what you are trying to do.

**Comment by jimboca** [comment]
Thanks, but that doesn't work for me because the ISY node servers are
kicked off in a thread, I'll keep looking into it.

**Comment by jimboca** [comment]
I was able to get this working in from inside the thread by getting the
asyncio loop from the main process and making that the current toop and
passing it to Elk constructor as loop.

Also, I realized after watching what I get in callback's that I can wait
until I get 'name' in the changeset to know the Area has been initialized.
Hopefully that works for Zones and everything else.

Thanks again for your help.
